### ISSUE TITLE
[2021 ICLR Oral] An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#53]
ViT 논문

CNN 없이 pure transformer만 사용하여 sota를 찍음

ViT, experiment, appendix만 설명

### 1.1. Model Overview
<img width="1327" alt="image" src="https://user-images.githubusercontent.com/49643709/154391202-458b101f-f0d6-4c6a-8da1-eed161b73f2c.png">
<img width="1242" alt="image" src="https://user-images.githubusercontent.com/49643709/154391666-dc0a3235-c410-4f05-99f9-9658c5a7f228.png">
<img width="1084" alt="image" src="https://user-images.githubusercontent.com/49643709/154411429-d9331991-8187-4c8c-ae57-df195ecab518.png">

1. (H, W, C) 이미지를 N개의 (P, P, C) 패치로 나눔
2. 패치를 flatten한 후, D dimension으로 linear projection (1 FC layer)
3. BERT와 같이 learnable class token을 patch embeddings 앞에 추가
4. Patch embeddings에 standard learnable 1D position embeddings를 더해줌
5. Transformer encoder를 거쳐 image representation y를 얻음
6. Classification head는 pre-training할 때에는 2 FC, fine-tuning할 때에는 1 FC 사용

### 1.2. Inductive Bias

CNN은 locality, 2-dimensional neighborhood structure, translation equivariance한 inductive bias를 가지고 있음

ViT는 MLP layer가 local, translationally equivariant한 inductive bias를 제외하고 별다른 inductive bias가 없음
(self-attention layer는 global하기에)

Image를 patch로 나누는 것과 resolution이 다를 때 position embedding을 조절한다는 점을 2-dimensional neighborhood structure이기에 inductive bias로 볼 수 있으나, 매우 약함

### 1.3. Hybrid Architecture

Patch embedding을 만들 때 raw image patch대신, CNN으로 patch feature map을 사용한 architecture

### 1.4. Fine-tuning and Higher Resolution

Fine-tune할 때 pre-trained prediction head를 제거하고 zero-initialized 1 FC layer를 붙임

Fine-tune할 때 pre-training보다 더 큰 resolution을 사용

Resolution이 커졌지만, patch size는 그대로 두어 더 긴 sequence length를 가지게 됨

이렇게 되면 pre-trained position embedding이 차원이 맞지 않아 2D interpolation을 통해 해결

### 2.1. Model Variants
<img width="922" alt="image" src="https://user-images.githubusercontent.com/49643709/154616418-1becc95b-9e3e-48ae-95df-5984368d6d15.png">

Model size와 input patch size를 이용하여 모델명을 생성
(ViT-L/16 : ViT Large model with 16 x 16 input patch size)

### 2.2. Comparison to State of the Art
<img width="1251" alt="image" src="https://user-images.githubusercontent.com/49643709/154617515-7a13bc43-5378-48d5-b9f4-88829e34f854.png">
<img width="1385" alt="image" src="https://user-images.githubusercontent.com/49643709/154894991-82257153-13d2-49d5-b413-436948af2c7e.png">

<img width="1214" alt="image" src="https://user-images.githubusercontent.com/49643709/154647223-a05de596-8ce6-4e44-b882-02e77b385483.png">

JFT-300M pre-trained ViT-L/16이 BiT-L보다 학습하는데 computational resource가 덜 들면서 성능이 좋음

더 큰 모델인 ViT-H/14는 대부분의 dataset에서 sota를 찍음

### 2.3. Pre-training Data Requirements

Dataset size가 얼마나 중요한지 실험

#### 2.3.1. ImageNet, ImageNet-21K, JFT-300M
<img width="1202" alt="image" src="https://user-images.githubusercontent.com/49643709/154892048-5d430e64-c6b6-4286-b3a2-e6803e939b39.png">

<img src='https://user-images.githubusercontent.com/49643709/154888285-4a27bc7f-a1a8-48ae-95dd-af0d77f71d40.png' width=60% height=60%>

ViT 모델을 ImageNet, ImageNet-21K, JFT-300M 순서대로 pre-train

모델 학습을 돕기 위해 weight decay, dropout, label smoothing 총 3가지 regularization 사용
(Dropout의 경우 qkv projection을 제외한 모든 fc layer 뒤에 적용)

ImageNet에서 학습 시 regularization을 사용해도 ViT-Large 모델이 ViT-Base 모델보다 성능이 낮음

ImageNet-21K에서 학습 시 ViT-Large 모델과 ViT-Base 모델의 성능이 비슷함

JFT-300M에서 학습 시 ViT-Large 모델이 ViT-Base 모델보다 성능이 좋음

ViT 모델은 ImageNet에서 BiT (CNN) 모델보다 성능이 낮지만 더 큰 데이터셋에서 성능이 좋음

#### 2.3.2. Subsets of JFT-300M
<img src='https://user-images.githubusercontent.com/49643709/154888316-235b211f-a173-45f0-9d7d-d3fd11710075.png' width=60% height=60%>

JFT-300M에서 9M, 30M, 90M의 random subsets를 만듬

모든 세팅에서 똑같은 hyperparameter 사용하고 regularization 없이 학습하여 model의 intrinsic property를 확인

결과를 통해 convolution의 inductive bias가 smaller dataset에서 유용하다는 것을 확인할 수 있음

하지만 larger dataset에서는 data로 relevant pattern을 학습하는 것이 더 유익함

### 2.4. Scaling Study
<img width="1236" alt="image" src="https://user-images.githubusercontent.com/49643709/154892082-ec7e054f-4be8-4f1e-9045-ea0ba0619764.png">

<img width="1189" alt="image" src="https://user-images.githubusercontent.com/49643709/154888584-9d6275c7-135e-4dd8-88ef-72c6821d5981.png">

Pre-train dataset을 JFT-300M으로 고정해두어 모델 성능에 bottleneck이 없게 세팅

몇가지 유의미한 패턴이 관측됨

1. Vision Transformer는 performance/compute trade-off에서 측면에서 ResNet보다 좋음
2. Hybrid 모델은 small computation budget에서 ViT보다 성능이 더 좋지만 모델이 커지면 성능 차이가 사라짐
3. Vision Transformer로 학습 시 saturate되는 현상이 관측되지 않음

### 2.5. Inspecting Vision Transformer
<img width="1203" alt="image" src="https://user-images.githubusercontent.com/49643709/154889516-33871f98-612e-4a76-a534-4b8f0ac8ff97.png">

<img width="1206" alt="image" src="https://user-images.githubusercontent.com/49643709/154894740-8271cb5f-0daa-4172-b8f3-a88035ba3f83.png">

Vision Transformer가 어떻게 작동하는지 이해하기 위해 다양한 분석 진행

1. Internal representation 분석
    - Flatten patch에서 embedding을 만드는 learned embedded filters 중 top principal component만 시각화한 결과가 fine structure의 low-dimensional representation을 위한 basis function과 매우 유사함
2. Position embedding 분석
    - Position embedding간의 유사도를 보면 이미지 내의 distance를 학습하는 것을 확인할 수 있음
    - 같은 row, column에 해당하는 경우 embedding 유사도가 높아 row-column structure를 학습함음 확인할 수 있음
    - Larger grid에서 sinusoidal structure 패턴이 등장
3. Average distance 분석
    - 각 layer별, head마다 attention weight간의 average distance를 측정하여 receptive field가 어떻게 되는지 확인
    - 몇몇 head는 이미 lowest layer에서 image 전반을 다 보고 있음을 확인할 수 있으며, 이는 model이 information을 globally하게 통합할 수 있음을 보여줌
    - Hybrid model의 경우 highly localized attention을 보기 어려운데, 해당 attention은 CNN의 early convolution layer과 비슷한 역할을 한다고 볼 수 있음
    - Network의 depth가 깊어질수록 attention distance가 증가하는데 이는 model이 classification과 semantically 관련된 image region을 globally 본다는 것을 알 수 있음

### 2.6. Self-Supervision
<img width="1114" alt="image" src="https://user-images.githubusercontent.com/49643709/154896014-865b45a6-898b-4a64-9845-34c4bbd11137.png">

Masked patch prediction task를 이용하여 self-supervised learning으로 학습

Scratch로 학습하는 것보다 성능이 더 좋았지만 supervised pre-training에 비해선 역부족

### 3.1. SGD vs. Adam for ResNets
<img width="896" alt="image" src="https://user-images.githubusercontent.com/49643709/154892607-2e721483-0bd3-4386-a271-7a87521b3733.png">

SGD, Adam으로 JFT pre-train한 ResNet (50x1, 152x2)에 대해 fine-tune performance 결과를 비교

Adam으로 pre-train한 결과가 SGD보다 대부분의 dataset에서 성능이 좋음

### 3.2. Transformer Shape
<img width="1144" alt="image" src="https://user-images.githubusercontent.com/49643709/154892811-27e04056-f7fc-498b-86a4-511229f07f43.png">

Depth를 scaling하면 효과가 가장 크지만 한계가 있음

Width를 scaling하는 효과가 가장 미미함

Patch size를 줄이고 effective sequence length를 증가하면 효과가 있었음

### 3.3. Head type and Class token
<img width="1242" alt="image" src="https://user-images.githubusercontent.com/49643709/154893357-4a857237-96a5-4519-b7ba-6aa788cc2e5b.png">

Class token을 사용하지 않고 image-patch embedding한 결과를 GAP하여 분류해도 충분히 성능이 괜찮음

다만 class token을 사용한 모델과 다른 학습 방법을 사용해야함

### 3.4. Positional Embeddings
<img width="1203" alt="image" src="https://user-images.githubusercontent.com/49643709/154893392-60beee90-5137-44a7-8521-15fc874e2654.png">
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/49643709/154894096-968e418b-bc35-46b2-8b47-9aa38a2eb52a.png">

Positional embedding을 넣은 것과 안넣은 결과의 성능 차이가 크기에 positional embedding은 넣어야함

Position information을 어떻게 넣어주는지는 pixel-level이 아닌 patch-level input이기에 spatial relation을 학습하는 것은 쉬워 중요하지 않고 성능 차이도 미미함

다만 training hyperparameter에 따라 position embedding의 similarity pattern이 조금씩 다름

### 3.5. Empirical Computational Costs
<img width="1184" alt="image" src="https://user-images.githubusercontent.com/49643709/154894151-96d97d54-f009-4135-ad20-46d458af8929.png">

ViT가 ResNet보다 더 memory-efficient함

### 3.6. Axial Attention
<img width="1217" alt="image" src="https://user-images.githubusercontent.com/49643709/154894574-c3fe4f75-c92f-4cca-9fa2-60cf0281d8ae.png">

Axial-ViT-B/32, Axial-ViT-B/16이 ViT-B보다 성능은 좋지만 compute cost가 큼

### 3.7. Attention Maps
<img width="452" alt="image" src="https://user-images.githubusercontent.com/49643709/154895055-b8bde2fa-4d85-4c4c-a1a2-190c31ef1463.png">

Attention Rollout을 통해 output token에서 input token으로의 attention map을 구함
(ViT-L/16 모델에 대해 attention weight를 average하여 모든 layer에 대해 weight matrix를 recursively multiply)

**Comment by Jasonlee1995** [comment]
참고로 average distance에서 나오는 attention weight는 softmax(query dot key) 부분을 의미함

즉 average of distance(M, N) * softmax(query_M dot key_N)임
